### ISSUE TITLE
SRW - Partially Coherent Simulation Freezes At its Time Limit

### ISSUE THREAD

**Issue by DmitrySmalyuk** [radiasoft/sirepo#267]
I started a partially coherent beamline simulation on SRW on expdev.nsls2.bnl.gov server, where the time limit for partially coherent simulations should be three days. After more than three days, the interface displayed that the simulation was still running, and showed an elapsed time value of 2 days and 23:59:48, or 12 seconds short of three days. I left it to go on, but even after several hours, it still displayed the same elapsed time value - 12 seconds short of its time limit. It seems like the simulation froze right before it was supposed to stop. The report that the simulation was producing did not seem to be changing at all after the three days passed. A screenshot is attached.

![bugwiththreedaypartcoherentsims](https://cloud.githubusercontent.com/assets/20563531/17494774/2f938d72-5d84-11e6-8b30-723fc14ba6e0.jpg)

**Comment by robnagler** [comment]
@DmitrySmalyuk I'm working on background execution right now. 

Would you happen to have any server log output? There might be a background.log file in the run directory: `find ~/src/radiasoft/sirepo/run -name background.log`

**Comment by mrakitin** [comment]
Hi Rob, I checked the server (cpu-001), and we don't have any background.log in `/var/db/`. Any other places to look at?

**Comment by robnagler** [comment]
I thought it was a local run, because it was 3 days. background.log is only for local runs. Check the celery and sirepo log for cpu-001.

**Comment by mrakitin** [comment]
I have a bunch of such errors in `app.log`:
```python
[2016-08-08 13:04:41,862] ERROR in app: Exception on /simulation-frame/srw*5xh8so7z*multiElectronAnimation**2*1470429746 [GET]
Traceback (most recent call last):
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/server.py", line 469, in app_simulation_frame
    response = flask.jsonify(template.get_simulation_frame(run_dir, data, model_data))
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/template/srw.py", line 504, in get_simulation_frame
    return extract_report_data(str(run_dir.join(_MULTI_ELECTRON_FILENAME_FOR_MODEL[data['report']])), data)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/template/srw.py", line 153, in extract_report_data
    info = _remap_3d(info, allrange, file_info[filename][0][3], file_info[filename][1][2])
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/template/srw.py", line 1254, in _remap_3d
    auxAr = np.array('d', [0]*lenAr2d)
TypeError: data type not understood
```

Also there are some messages like:
```python
[2016-08-08 13:16:48,980] ERROR in app: Exception on /simulation-frame/srw*5xh8so7z*multiElectronAnimation**26*1470429746 [GET]
Traceback (most recent call last):
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/server.py", line 469, in app_simulation_frame
    response = flask.jsonify(template.get_simulation_frame(run_dir, data, model_data))
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/template/srw.py", line 504, in get_simulation_frame
    return extract_report_data(str(run_dir.join(_MULTI_ELECTRON_FILENAME_FOR_MODEL[data['report']])), data)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/sirepo-20160801.223635-py2.7.egg/sirepo/template/srw.py", line 137, in extract_report_data
    data, mode, allrange, arLabels, arUnits = uti_plot_com.file_load(filename)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_plot_com.py", line 116, in file_load
    return _file_load(fname, read_labels)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_plot_com.py", line 71, in _file_load
    data = uti_io.read_ascii_data_cols(_fname, '\t', _i_col_start=0, _i_col_end=0, _n_line_skip=nLinesHead)[0]
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_io.py", line 48, in read_ascii_data_cols
    resCols[colCountTrue].append(float(curPart))
ValueError: invalid literal for float(): 2.386553338855424e
```

**Comment by robnagler** [comment]
Do you know if these relate to the simulation Dmitry was working on?

**Comment by mrakitin** [comment]
I'm not sure if this relate to his calculations, but approximately that time Dmitry checked the status of his calculations.

**Comment by robnagler** [comment]
Can you check his IP and/or User Agent string? Would be good to get the
exact errors. Celery/Sirepo should catch all errors of this kind.

**Comment by mrakitin** [comment]
Yes, I'll do it tomorrow, Dmitry already left today. Do I need to grep his IP in the Celery/Sirepo logs?

**Comment by robnagler** [comment]
Yes, we want to figure out what happened on the server at the time of the
error.

**Comment by DmitrySmalyuk** [comment]
User Agent String:
```
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36
```
IP adress: `130.199.222.137`

**Comment by mrakitin** [comment]
Rob, I don't think we can gather the information using the IP address since access to cpu-001 is proxied, and in the log I see just the address `10.0.137.37` for all requests. What can I do with the user agent string?

**Comment by robnagler** [comment]
We may not write that in the uwsgi log, since we expect the proxy to write
it...

The proxy should pass the IP via an X-Forwarded-For or something.

Not sure we can do anything at this point. I do know that if the simulation
crashes, the UI doesn't report it always. Let me get the new code in place
(should be done this week). I'll try to add more logging, too.

**Comment by mrakitin** [comment]
OK, sounds good.

**Comment by mrakitin** [comment]
I asked admins to configure `X-Forwarded-For`.

**Comment by robnagler** [comment]
You might want all these:

```text
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Via $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
```

Don't forget to have them block the [HTTP Poxy](https://httpoxy.org/) attack with (nginx):

```text
```

**Comment by mrakitin** [comment]
Thanks Rob!

**Comment by robnagler** [comment]
Just encountered another conversion error:

```text
Overwrapped regexp I-search: lineparts
10.10.10.1 - - [17/Aug/2016 22:56:16] "GET /simulation-frame/srw*ybLsEhbx*multiElectronAnimation**2*1471474542 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/beaker/middleware.py", line 155, in __call__
    return self.wrap_app(environ, session_start_response)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/server.py", line 413, in app_simulation_frame
    response = _json_response(template.get_simulation_frame(run_dir, data, model_data))
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/srw.py", line 487, in get_simulation_frame
    return extract_report_data(str(run_dir.join(_MULTI_ELECTRON_FILENAME_FOR_MODEL[data['report']])), data)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/srw.py", line 136, in extract_report_data
    data, mode, allrange, arLabels, arUnits = uti_plot_com.file_load(filename)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_plot_com.py", line 116, in file_load
    return _file_load(fname, read_labels)
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_plot_com.py", line 71, in _file_load
    data = uti_io.read_ascii_data_cols(_fname, '\t', _i_col_start=0, _i_col_end=0, _n_line_skip=nLinesHead)[0]
  File "/home/vagrant/.pyenv/versions/2.7.10/lib/python2.7/site-packages/uti_io.py", line 48, in read_ascii_data_cols
    resCols[colCountTrue].append(float(curPart))
ValueError: could not convert string to float:
```

**Comment by robnagler** [comment]
Added timeout for requests, which should help with this case. Fix in 8e7b244
